pkg/server/middleware/request_limits.go

```diff
@@ -29,12 +40,24 @@ func (l RequestLimits) Wrap(next http.Handler) http.Handler {
 		body, err := io.ReadAll(reader)
 		if err != nil {
 			level.Warn(log).Log("msg", "failed to read request body", "err", err)
-			http.Error(w, fmt.Sprintf("failed to read request body: %v", err), http.StatusInternalServerError)
-			return
+			_ = level.Warn(l.logger).Log("msg", "middleware.RequestLimits.Wrap failed to read request body", "err", err)
```

Comment:
This is already being logged in the line above. Or are there any reasons to use the struct's logger instead of the span's from here?

Reply:
If I'm not mistaken spanlogger adds logs inside the span. 
So if the tracing is not working properly and spans are not sent we will not be able to see these logs (This is what happening in the faro collector right now, not all spans are sent).
That's why I added a logger here, to at least see these logs in Loki when I can't see them in Tempo

Reply:
I'm not a tracing expert, but I think the logger from `spanlogger` logs like a regular logger. It [combines](https://github.com/grafana/dskit/blob/c30720760c5d23eef733a3e7e299e8247e438a24/spanlogger/spanlogger.go#L34) a span and a logger, and adds the traceID as a field. And the things [we log](https://github.com/grafana/transmog/blob/94f5dc4893a9ff08a2686116bead85019c09b2f5/pkg/graphite/querier/api/render.go#L65) we see them in [Loki](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22graphite-querier%5C%22%7D%7C%3D%20%20%60render%20request%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1).

But I may be missing something here. Did you see this not working as expected?

Reply:
👋 - I never hit "Submit" on my review, my take on this was that spanlogger is a utility that adds `elapsed` to the log line, rather than having anything to do with tracing (I saw "mimir/utils/spanlogger" and thought it must be a utility they've written to simplify calculating the "span" of time, but I could be wrong and it could be doing something with trace spans ).

Reply:
I think that apart from that it's main purpose is to unify the creation and holding of both a span and a logger. That's why it's always followed by `defer log.Span.Finish()`

```go
log, ctx := spanlogger.New(r.Context(), "middleware.RequestLimits.Wrap")
defer log.Span.Finish()
```

Reply:
> And the things [we log](https://github.com/grafana/transmog/blob/94f5dc4893a9ff08a2686116bead85019c09b2f5/pkg/graphite/querier/api/render.go#L65) we see them in [Loki](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22graphite-querier%5C%22%7D%7C%3D%20%20%60render%20request%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1).

Thank you for the example! Trying to find faro-collector logs the same way you do for graphite proxy, but I don't see logs from spanlogger, only logs from struct's logger. Need to debug and understand why it happens
[Example of faro-collector logs](https://ops.grafana-ops.net/explore?panes=%7B%22NrH%22:%7B%22datasource%22:%22000000193%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bname%3D%5C%22faro-collector%5C%22%7D%20%7C%3D%20%20%60failed%20to%20read%20request%20body%60%20%21%3D%20%5C%22middleware.RequestLimits.Wrap%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22000000193%22%7D,%22editorMode%22:%22code%22%7D%5D,%22range%22:%7B%22from%22:%221689768047678%22,%22to%22:%221689768376397%22%7D%7D%7D&schemaVersion=1&orgId=1) (use query which filter out struct's log, they have prefix `middleware.RequestLimits.Wrap`)

Reply:
@mar4uk - I think I know what's going on.
```
level.Warn(log).Log("msg", "failed to read request body", "err", err)
```
^ This was added in a previous PR, and that `log` isn't logging anywhere. It wasn't connected to the Faro tracing issues, etc, I think it's just not going to be visible unless `l.logger` is used.

I think one-or-other of the `level.Warn(...).Log(...)` lines will work, but the using the spanlogger to give us the timings would be pretty useful when it's a timeout in middleware we're reporting.

Reply:
Graphite Proxies does this:
```
log.Logger = app.Logger
```
That log.Logger is from: "github.com/grafana/mimir/pkg/util/log"

```
var (
	// Logger is a shared go-kit logger.
	// TODO: Change all components to take a non-global logger via their constructors.
	// Prefer accepting a non-global logger as an argument.
	Logger = log.NewNopLogger()
)
```
So I'll see how our logging can be integrated into this mode of initialization, and we can forego passing a logger as an arg to middleware constructors 👍

Reply:
Thanks @npazosmendez - I discussed offline with @mar4uk and I've a little more free time right now, so I'll push a commit asap to remove the logger Faro was using, and we'll test Faro works via changes to how we initialize the context in our own time.
Thanks for the suggestion

Reply:
@npazosmendez @jesusvazquez - I think this is resolved now. We've kept the spanlogger, and separately I will look at how to get Faro logging setup to use it (by copying the example from graphite-proxy, most likely).
But this PR is now ready for general consumption - thanks @mar4uk for doing the hardest part! (and thank you @jesusvazquez and @npazosmendez for the equally hard reviewing part!). 
🙇